src/datumaro/components/importer.py

```diff
@@ -165,3 +170,105 @@ def get_extractor_merger(self) -> Optional[Type[ExtractorMerger]]:
             Otherwise, use the return type to merge the extractors.
         """
         return None
+
+
+def with_subset_dirs(input_cls: Type[Importer]):
+    """
+    Transforms an importer that can parse just 1 subset into one that can parse several nested
+    directories as subsets.
+
+    This decorator only recognizes directories from SUBSET_NAME_WHITELIST as candidate subsets.
+
+    Example:
+
+    Suppose there is a directory structure like this:
+
+    .. code-block::
+
+        root/
+        ├── train/
+        │   └── ....
+        └── valid/
+            └── ....
+
+    and we have FooImporter that can parse datasets either from root/train/ or root/valid/
+    dirs, but not both at once. with_subset_dirs() decorator can be used to define
+    a new Importer that can parse both subset directories from the root directory:
+
+    .. code-block::
+
+        @with_subset_dirs
+        class FooWithSubsetDirsImporter(FooImporter):
+            pass
+
+    Now, FooWithSubsetDirsImporter can be used to import a combined dataset from root folder.
+    """
+
+    @wraps(input_cls, updated=())
+    class WrappedImporter(input_cls):
+        NAME = input_cls.NAME
+
+        @classmethod
+        def detect(
+            cls,
+            context: FormatDetectionContext,
+        ) -> Optional[FormatDetectionConfidence]:
+            @contextmanager
+            def _changed_context_root_path(context: FormatDetectionContext, path: str):
+                tmp = context.root_path
+                context._root_path = path
+
+                try:
+                    yield
+                finally:
+                    context._root_path = tmp
+
+            confs = []
+            path = Path(context.root_path)
+
+            if not path.is_dir():
+                context.fail(
+                    f"{input_cls.NAME} should require an input as a directory path. "
+                    f"However, {path} is not a directory path."
+                )
+
+            for sub_path in path.iterdir():
+                if sub_path.name.lower() not in SUBSET_NAME_WHITELIST or not sub_path.is_dir():
+                    continue
+
+                with _changed_context_root_path(context, str(sub_path)):
+                    conf = input_cls.detect(context)
+                if conf is not None:
+                    confs.append(conf)
+
+            if not confs:
+                context.fail(f"{input_cls.NAME} cannot find its subdirectory structure.")
+
+            return max(confs)
```

Comment:
I can see in many importers we have None as return value, but it doesn't match the function prototype (it's supposed to always return confidence). Could you check this and make sure the `detect()` method is defined correspondingly to the function in the base class?